fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/InstrumentedTransaction.java

```diff
@@ -0,0 +1,401 @@
+import com.apple.foundationdb.KeySelector;
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.MutationType;
+import com.apple.foundationdb.Range;
+import com.apple.foundationdb.ReadTransaction;
+import com.apple.foundationdb.StreamingMode;
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.TransactionOptions;
+import com.apple.foundationdb.async.AsyncIterable;
+import com.apple.foundationdb.async.AsyncIterator;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.provider.common.StoreTimer;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Wrapper around {@link Transaction} that instruments certain calls to expose their behavior with
+ * {@link FDBStoreTimer} metrics.
+ */
+public class InstrumentedTransaction implements Transaction {
+
+    @Nullable
+    protected StoreTimer timer;
+    protected Transaction underlying;
+
+    public InstrumentedTransaction(@Nullable StoreTimer timer, Transaction underlying) {
+        this.timer = timer;
+        this.underlying = underlying;
+    }
+
+    @Override
+    public void addReadConflictRange(byte[] keyBegin, byte[] keyEnd) {
+        underlying.addReadConflictRange(keyBegin, keyEnd);
+    }
+
+    @Override
+    public void addReadConflictKey(byte[] key) {
+        underlying.addReadConflictKey(key);
+    }
+
+    @Override
+    public void addWriteConflictRange(byte[] keyBegin, byte[] keyEnd) {
+        underlying.addWriteConflictRange(keyBegin, keyEnd);
+    }
+
+    @Override
+    public void addWriteConflictKey(byte[] key) {
+        underlying.addWriteConflictKey(key);
+    }
+
+    @Override
+    public void set(byte[] key, byte[] value) {
+        underlying.set(key, value);
+        increment(FDBStoreTimer.Counts.WRITES, FDBStoreTimer.Counts.BYTES_WRITTEN, key.length + value.length);
+    }
+
+    @Override
+    public void clear(byte[] key) {
+        underlying.clear(key);
+        increment(FDBStoreTimer.Counts.DELETES);
+    }
+
+    @Override
+    public void clear(byte[] keyBegin, byte[] keyEnd) {
+        underlying.clear(keyBegin, keyEnd);
+        increment(FDBStoreTimer.Counts.DELETES);
+    }
+
+    @Override
+    public void clear(Range range) {
+        underlying.clear(range);
+        increment(FDBStoreTimer.Counts.DELETES);
+    }
+
+    @Override
+    @Deprecated
+    public void clearRangeStartsWith(byte[] prefix) {
+        underlying.clearRangeStartsWith(prefix);
+        increment(FDBStoreTimer.Counts.DELETES);
+    }
+
+    @Override
+    public void mutate(MutationType opType, byte[] key, byte[] param) {
+        underlying.mutate(opType, key, param);
+        /* Do we want to track each mutation type separately as well? */
```

Comment:
As noted, do we want per-mutation type counters as well? Obviously can always come later...

Reply:
I think this is probably fine for now. Maybe some day, though.